### PR TITLE
Code salvage: Centcomm orders

### PR DESCRIPTION
**Pull request by MadmanMartian**
while I was looking at the supply_shuttle code, I noticed something called 'centcomm orders'

They're basically custom orders that centcomm can request of the station. Going to try and salvage this and make it into a random event

- [x] Centcomm orders work again
- [x] Active orders can be browsed in some way
- [x] Don't break the supply shuttle
- [x] Populate the list of things that can be requested
- [x] Balance Tweaks

### CODE REVIEW

File: code/modules/events/centcomm_order.dm [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+					"amount" = rand(1,5),
+					"value" = rand(350,1250),
+					),
+				list(
+					"item" = /obj/item/weapon/subspacetunneler,
+					"amount" = rand(1,3),
+					"value" = rand(350,1250),
+					)
+				)
+		if("Science")
```
Comment (Arthurdentist) [inline code/modules/events/centcomm_order.dm]: Shouldn't this be if("Medical")? looks like you mixed science and medical up by mistake.

File: code/modules/events/centcomm_order.dm [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+					"item" = /obj/item/stack/sheet/mineral/clown,
+					"amount" = rand(1,10),
+					"value" = rand(200, 3500)
+					)
+				)
+
+		if("Medical")
+			chosen = pick(
+				list(
+					"item" = /obj/item/weapon/gun/energy/gun/nuclear,
```
Comment (Kurfursten) [inline code/modules/events/centcomm_order.dm]: Pretty sure this is an AEG, and science only gets modkits now

File: code/modules/events/centcomm_order.dm (hunk 1 above)
Comment (MadmanMartian) [reply code/modules/events/centcomm_order.dm]: Coding at 2AM'll do that to ye

File: code/modules/events/centcomm_order.dm (hunk 2 above)
Comment (MadmanMartian) [reply code/modules/events/centcomm_order.dm]: Then science will have to get energy guns to upgrade from somewhere

Comment (Kurfursten) [reply code/modules/events/centcomm_order.dm]: I just don't see it happening that science, cargo, and security all team up and sacrifice one of the few eguns on the station for, at most, 1250 credits.

Comment (Pathid) [reply code/modules/events/centcomm_order.dm]: Can't cargo order a crate that has 2 eguns for 50 credits? I don't get what the issue is.

File: code/game/centcomm_orders.dm [hunk 3]
```diff
@@ -34,12 +31,16 @@ var/global/current_centcomm_order_id=124901
...
+		if(istype(O, /obj/item/stack))
+			var/obj/item/stack/S = O
+			amount = S.amount
 		if(!(O.type in fulfilled))
 			fulfilled[O.type]=0
 		// Don't claim stuff that other orders may want.
 		if(fulfilled[O.type]==requested[O.type])
 			return 0
-		fulfilled[O.type]=fulfilled[O.type]+1
+		fulfilled[O.type]=fulfilled[O.type]+amount
```
Comment (PJB3005) [inline code/game/centcomm_orders.dm]: Use `+=`

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by MadmanMartian** [comment]
supply_console2_1
supply_console2_2

File: code/game/machinery/computer/cargo.dm [hunk 4]
```diff
@@ -333,6 +337,11 @@ For vending packs, see vending_packs.dm*/
 			return
 		supply_shuttle.requisition = text2num(href_list["requisition_status"])
 		return 1
+	else if (href_list["screen"])
+		if(!check_restriction(usr))
+			return
+		screen = text2num(href_list["screen"])
```
Comment (PJB3005) [inline code/game/machinery/computer/cargo.dm]: Sanitize this so it can only be `SRC_MAIN` or `SRC_CENTCOM`

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: code/game/machinery/computer/cargo.dm [hunk 5]
```diff
@@ -340,7 +340,8 @@ For vending packs, see vending_packs.dm*/
 	else if (href_list["screen"])
 		if(!check_restriction(usr))
 			return
-		screen = text2num(href_list["screen"])
+		if(text2num(href_list["screen"]) == SCR_MAIN || text2num(href_list["screen"]) == SCR_CENTCOM)
```
Comment (PJB3005) [inline code/game/machinery/computer/cargo.dm]: Store the value of this `text2num` instead of running it 3 times.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by SonixApache** [comment]
when

**Comment by MadmanMartian** [comment]
now that #16785 is merged, I've been able to test and find that centcom orders aren't added to the supply consoles.

**Comment by MadmanMartian** [comment]
fixed it

File: code/game/supplyshuttle.dm [hunk 6]
```diff
@@ -298,4 +299,24 @@ var/datum/controller/supply_shuttle/supply_shuttle = new
...
 			shoppinglist += O
 		else
-			to_chat(user, "<span class='warning'>[O.orderedby] does not have enough funds for this request.</span>")
\ No newline at end of file
+			to_chat(user, "<span class='warning'>[O.orderedby] does not have enough funds for this request.</span>")
+
+/datum/controller/supply_shuttle/proc/add_centcomm_order(var/datum/centcomm_order/C)
+	centcomm_orders.Add(C)
+	for(var/I in centcomm_orders)
+		to_chat(world, "[I]")
```
Comment (DamianX) [inline code/game/supplyshuttle.dm]: worse than hitler

File: code/game/centcomm_orders.dm [hunk 7]
```diff
@@ -52,6 +53,40 @@ var/global/current_centcomm_order_id=124901
...
+/datum/centcomm_order/proc/getFulfilledByName(var/html_format = 0)
+	var/manifest = ""
+	if(html_format)
+		manifest = "<ul>"
+	for(var/path in fulfilled)
+		if(!path)
+			continue
+		var/atom/movable/AM = new path()
+		if(html_format)
+			manifest += "<li>[AM.name], amount: [fulfilled[path]]</li>"
```
Comment (DamianX) [inline code/game/centcomm_orders.dm]: can't you use `initial(path.name)` to avoid creating these dummy objects?

**Review by DamianX — CHANGES_REQUESTED** [review]

File: code/game/supplyshuttle.dm (hunk 6 above)
Comment (MadmanMartian) [reply code/game/supplyshuttle.dm]: Absolutely adolf

File: code/game/centcomm_orders.dm (hunk 7 above)
Comment (MadmanMartian) [reply code/game/centcomm_orders.dm]: It doesn't work as it needs an object to find the initial value of that variable for.

File: code/game/centcomm_orders.dm [hunk 8]
```diff
@@ -52,6 +53,40 @@ var/global/current_centcomm_order_id=124901
...
+		manifest = "<ul>"
+	for(var/path in fulfilled)
+		if(!path)
+			continue
+		var/atom/movable/AM = path
+		if(html_format)
+			manifest += "<li>[initial(AM.name)], amount: [fulfilled[path]]</li>"
+		else
+			manifest += "[initial(AM.name)], amount: [fulfilled[path]]"
+		qdel(AM)//just to make sure they're deleted by the garbage collector
```
Comment (DamianX) [inline code/game/centcomm_orders.dm]: AM is never instantiated with the initial business, this isn't necessary

File: code/game/centcomm_orders.dm [hunk 9]
```diff
@@ -52,6 +53,40 @@ var/global/current_centcomm_order_id=124901
...
 	acct.charge(-worth,null,"Payment for order #[id]",dest_name = name)
 
+/datum/centcomm_order/proc/getRequestsByName(var/html_format = 0)
+	var/manifest = ""
+	if(html_format)
+		manifest = "<ul>"
+	for(var/path in requested)
+		if(!path)
+			continue
+		var/atom/movable/AM = new path()
```
Comment (DamianX) [inline code/game/centcomm_orders.dm]: the same business applies here

**Review by DamianX — CHANGES_REQUESTED** [review]

**Comment by Hugh-G-Rection** [comment]
>Removes Hitler

:ok_hand:

**Comment by MadmanMartian** [comment]
@PJB3005 @DamianX changes made

**Review by PJB3005 — APPROVED** [review]

**Review by DamianX — APPROVED** [review]

**Pull request merged**